helix-term/src/ui/popup.rs

```diff
@@ -277,23 +262,30 @@ impl<T: Component> Component for Popup<T> {
             .required_size((inner.width, inner.height))
             .expect("Component needs required_size implemented in order to be embedded in a popup");
 
-        self.child_size = (width, height);
-        self.size = (
+        let size = (
```

Comment:
If you look at thedDefinition of required size it says you should not be clamping the size to the size of the viewport. I would like to follow the documented definition of the function. The reason the old code was buggy and confusing is because people did adhox things that kind of worked but didn't follow the actual api.

Reply:
I think I understood the API very well. What this code is doing is that, it reads the content requested_size, which can  be infinite and clamp it to the popup maximum size, which IMO fine, because the popup shouldn't request higher size than it's maximum size

This PR didn't change anything regarding the requested size of the popup.  It only removes the mutation of `self.size` and `self.child_size` at the cost that requested_size will be called twice in the completion popup, for the other popup there should be no drawbacks.

if you don't like removing the mutation, maybe I can modify the PR to only improve the popup behavior without much modifications

Reply:
I quite like removing the mutation.  I think you got confused where I put the review comment (at the top of the function). I was talking about the following code:

```
            (width + self.margin.width()).min(max_width),
            (height + self.margin.height()).min(max_height),
```

The capping (`.min`) is what shouldn't happen in required size. That is the API contract. Instead you should do that wen calling `required_size` (I think in area)

Reply:
the code you mentioned was already there. check the changes :sweat_smile: , this PR didn't change anything regarding the output of `required_size`

> The capping (.min) is what shouldn't happen in required size. That is the API contract. Instead you should do that wen calling required_size (I think in area)

I disagree for the following reason: the maximum size of the popup is hard coded to (120,26). why should the popup's `required_size` returns any higher value than its maximum size? IMO the current (and old) implementation comply with the API description, unless you want to remove that hard coded maximum size and allow the Popup to have any size it likes.

Reply:
yes I understand that you didn't change this code but I am saying that the old code was wrong for doing this and you should fix it as part of this PR.

On second thaught I think the misunderstanding is coming from the fact that `required_size` just doens't make sense for popup it all. `required_size` is an *optinal* method of component so that containers can compute its size and add a scrollbar.

I was saying you shouldn't limit the size of the popup to the viewport here since that isn't the job of required_size (you want to return the true size so the parent add a scrollbar. Particularly I took issue with limiting it to the viewport size since the viewport size would be the size of the entire window, the true size computation depends on the cursor position and can't be done here. T

You are right that the width/height should be limited to the maximum popup width/height (120/26). I was not taking issue with that but the limiting to the viewport size which can messup calculations in area later (or is just useless most of the time).

I think we should just not implement `required_size` for popup at all. `required_size` really only make sense for components that go *inside a popup` and take up their whole parent (and are therefore cropped to the viewport). Instead, the logic should just be moved into the `area` function directly (that function does its own cropping anyway I think then it will be more clear why I took issue with that)

Reply:
sorry it took me a while to understand the issue :smiley: . basically those 2 lines does not follow the API description:

```
let max_width = 120.min(viewport.0);
let max_height = 26.min(viewport.1.saturating_sub(2)); // add some spacing in the viewport
```
you are right. Somehow I missed this. I think, this can be moved to the `area` function without any side effects or issues. What I don't like about the API is `mut self`. also in the `area` function there is `mut self`. but In this PR I am not going to make such huge changes to the API. I will keep it for another one :smile_cat: (If I find a better solution)

Reply:
I hope this solved/fine